### ISSUE TITLE
Slow, even with small dataset

### ISSUE THREAD

**Issue by ericcbohn** [Christopher-Thornton/hmni#3]
We built a 5000 row proof of concept that searches first and last name only, and it takes about a minute to show a result.  Our implementation would need to search a few million rows.  Do you have any suggestions on how to improve performance?

**Comment by Christopher-Thornton** [comment]
I have plans to further optimize the `Matcher` class and add optional parameters which will use heuristics to quickly filter down potential candidates.

**Comment by ericcbohn** [comment]
Batch processing was the first thing that came to my mind.  We just pulled your library down yesterday, so I think it's current.  Maybe we could help further things with you?  @odinolav

**Comment by Christopher-Thornton** [comment]
Thanks, I am very open to collaborating with other developers. The latest release is `v0.1.6` which I uploaded last night.

**Comment by odinolav** [comment]
For what it's worth, I'm first trying to speed up the batch test I have going, and I've found (unsurprisingly) that preemptively making sure at least two characters in any order match up can help quite a bit. At first I was thinking matching by two _consecutive_ characters could be safe... but then there are names like Jim => James. 2 non-consecutive seems safe though.

**Comment by Christopher-Thornton** [comment]
Agreed, I think one or more matching consonant characters would also be a safe candidate filter. I will have to run some tests on my international names dataset to confirm. If there are a few edge cases where it doesn't work, they can even be hard-coded into the algorithm.

**Comment by Christopher-Thornton** [comment]
I ended up using a modified version of the Soundex algorithm to filter for disjointed encodings. This candidate filter is a new feature in `v0.1.8` as default behavior `Matcher(prefilter=True)`. This alone will not make the unordered cartesian product of two 1M+ name sets computable, as the implementation is still O(m x n). I will be labelling this issue as help wanted for any optimization experts up for the challenge.
